backend/jobsearch/src/main/kotlin/com/jobsearch/service/VacancyService.kt

```diff
@@ -1,87 +1,111 @@
         }
+
+        val newVacancy = vacancyRepository.save(vacancyEntity)
+
+        return mapToVacancyResponseDto(newVacancy)
     }
 
-    fun updateVacancy(vacancyId: Int, vacancyDto: VacancyDto): VacancyDto {
-        val vacancy = vacancyRepository.findById(vacancyId)
-            .orElseThrow { NoSuchElementException("No vacancy found with id $vacancyId") }
+    @Transactional
+    fun updateVacancy(vacancyId: Int, vacancyDto: VacancyRequestDTO): VacancyResponseDTO {
 
-        val selectedJobFamily = jobFamilyService.findByJobFamilyId(vacancyDto.jobFamilyId!!)
-            .orElseThrow { NoSuchElementException("No vacancy found with id ${vacancyDto.jobFamilyId}") }
+        val vacancy = vacancyRepository.findById(vacancyId)
+            .orElseThrow { NotFoundException("No vacancy found with id $vacancyId") }
+        val manager = userService.retrieveAuthenticatedUser()
+        if (vacancy.manager.email != manager.email ) throw ForbiddenException("You are not allowed to edit this vacancy.")
 
+        val selectedJobFamily = jobFamilyService.findByJobFamilyId(vacancyDto.jobFamilyId)
         vacancy.name = vacancyDto.name
         vacancy.companyName = vacancyDto.companyName
         vacancy.salaryExpectation = vacancyDto.salaryExpectation
         vacancy.description = vacancyDto.description
         vacancy.jobFamily = selectedJobFamily
 
         val updatedVacancy = vacancyRepository.save(vacancy)
-        return mapToVacancyDto(updatedVacancy)
-    }
-
-    fun deleteVacancy(vacancyId: Int): String {
-        val vacancy = vacancyRepository.findById(vacancyId)
-            .orElseThrow { NoSuchElementException("No vacancy found with id $vacancyId") }
 
-        vacancyRepository.delete(vacancy)
-        return "Vacancy deleted successfully"
+        return mapToVacancyResponseDto(updatedVacancy)
     }
 
     @Transactional
-    fun findVacanciesByFilter(salary: Int?, jobFamilyId: Int?, yearsOfExperience: Int?): List<VacancyDto> {
-        val vacancies = vacancyRepository.findVacanciesByFilters(salary, jobFamilyId, yearsOfExperience)
-        return vacancies.map {
-            mapToVacancyDto(it)
-        }
-    }
-
-    fun mapToVacancyDto(vacancy: Vacancy): VacancyDto {
+    fun deleteVacancy(vacancyId: Int) {
+        val vacancy = vacancyRepository.findById(vacancyId)
+            .getOrElse { return }
+        val manager = userService.retrieveAuthenticatedUser()
+        if (vacancy.manager != manager ) throw ForbiddenException("You are not allowed to erase this vacancy.")
+        vacancyRepository.delete(vacancy)
+}
+
+    /**
+     * Maps a Vacancy object to a VacancyResponseDTO object.
+     *
+     * @param vacancy the Vacancy object to map
+     * @return the mapped VacancyResponseDTO object
+     */
+    fun mapToVacancyResponseDto(vacancy: Vacancy): VacancyResponseDTO {
         return vacancy.let {
-            VacancyDto(
-                it.id,
+            VacancyResponseDTO(
+                it.id!!,
```

Comment:
it is not a good practice to use `!!`
Please make this field not nullable or check is it is not null before using  it

Reply:
The mapper receives a vacancy created from the repository, so the id will always be non-null when this function is used.
The reason for id to be a nullable value is because when constructing the new entity to be saved in the database it will not have a defined id until it is returned by repository.save()